### PR TITLE
Fix duplication of some support events appearing in Productive

### PR DESCRIPTION
**Pull request by tahb**
Before this change, every time we ran the task, duplicate events appeared in Productive. It also appeared to make a lot more duplicates for past events, compared to future events.

We make a couple of changes to get this working as intended:

1. We grab more support events from Productive to match the 2 month history the Support rota returns by default. This prevents duplication as we are comparing the data sets that span the same time range. When things were in the rota but not in Productive, we kept creating them in Productive!
2. We stop creating and deleting **all** events from Productive on every run by making our 2 support objects comparable. Given the way `date` was a `date` in one version of `SupportRotaToProductive::SupportRotation` and `string` in another, they were always detected as different events. Our union and difference operations then failed to detect any matches and so deleted everything 

We then turn on the scheduled task as it now looks to be working after running it for real from my local machine several times.

### CODE REVIEW

**Comment by tahb** [comment]
I'm finding it really hard to update the test suite to accomodate this change. [An after hook is used to create booking stubs](https://github.com/dxw/support-rota-to-productive/blob/main/spec/factories/booking.rb#L94) everytime a factory is created. The hook has fixed filter date of `Date.today` which is no longer right. We need to inteogate the result of the support_rota to determine what this value should be.

The [import spec is full of mystery guests](https://github.com/dxw/breathe-to-productive/blob/main/spec/breathe_to_productive/import_spec.rb#L20) which makes it incredibly difficult to unpick what is really going on in each test in order to extend this behaviour. There are 16 `let` statements of which 8 are `let!`, as well as 2 `before` blocks which also do some stuff. This article better explains the difficulties of this approach: https://thoughtbot.com/blog/lets-not.

I will try to unpick this but I may need to start the testing approach again in order to figure it out.

**Comment by tahb** [comment]
I've redesigned a chunk of test suite so that I could make the fix by changing the way we need to query for Productive events. Everything is now green and we have encapsulated tests with a lot fewer `let` statements. The test suite is more repetitive in places but it is more readable and extendable.

File: lib/support_rota_to_productive/support_rotation.rb [hunk 1]
```diff
@@ -15,25 +15,37 @@ def hash
...
-        Productive::Booking.where(project_id: SUPPORT_PROJECT_ID, after: Date.today).all.map(&:to_support_rotation)
+        bookings = Productive::Booking
+          .where(project_id: SUPPORT_PROJECT_ID, after: date_of_earliest_support_event)
+          .all
+        bookings.map(&:to_support_rotation)
+      end
+
+      def date_of_earliest_support_event
+        first_support_date = from_support_rota.first
+        return Date.today unless first_support_date.respond_to?(:date)
```
Comment (evilstreak) [inline lib/support_rota_to_productive/support_rotation.rb:38]: Is the only circumstance that the object doesn't respond to `#date` when it's `nil`? I can't work out any other path to it.

If it's only for nil checking, would it be clearer to use something like `if from_support_rota.empty?`

**Review by evilstreak — APPROVED** [review]
The changes look reasonable. I think we'd be OK to ship it as is.

One question (besides one inline comment): Should we be going back 2 months to edit bookings?

I can see us wanting to go back maybe as far as a week, in case we get to the end of a week and are like "Oh actually, Tom did support this instead of Dom, let's reflect that in the rota" and then that makes a difference for their default options for time sheets.

But longer back than that, folks should already have recorded their time, and the bookings are no longer used. I'm not sure if modifying the bookings after the fact does any harm – it might be fine, but it's definitely not necessary.

So, rather than trying to gather extra Productive data to do the comparison, what if we dropped some of the support rota data instead? Or a combination of the two, if we want to go a few days into the past?

**Comment by evilstreak** [comment]
One more thought: would handling the dates in tests be simpler using timecop, which is already a dependency of this app?

**Comment by tahb** [comment]
Thanks for spending your time to go through this for me!

> Should we be going back 2 months to edit bookings?

Whilst useful to seed some more existing data I think we could explore this. Going back 1 week might work but it depends how and when corrections are made:

Last minute changes made due to sickness or unexpected leave may indeed make OpsGenie out of date after it has been pulled from on a given day. I suspect a correction would be made to OpsGenie during the morning of the date affected. If we don't go back at least one day in this task, by the time it next runs tomorrow, this correction would be ignored? It seems like we want at least 1 day.

The correction could also be made the next day, next week or next month. It's hard to know when or even if this would happen? This makes it arbitrary and without understanding who updates it and their drivers for doing so if might be irrelevant since OpsGenie may always be slightly wrong.

In all cases the person correcting the support rota would also have to remember to duplicate the change in Productive. This is inconvenient and more likely to be forgotten.

(I'll have a look into your comments on Timecop and guarding `nil` dates too)

File: lib/support_rota_to_productive/support_rotation.rb (hunk 1 above)
Comment (tahb) [reply lib/support_rota_to_productive/support_rotation.rb]: In the end this method has been removed so I took no further action.

**Comment by tahb** [comment]
I've changed the approach to filtering down the response from the support rota to 1 day ago. This would allow for corrections made on a given day to be updated by the next job that runs.

If the task does not run, or if the correction is made more than 1 day later then a manual correction in Productive would be needed. 

The decoupling between support rota responses and the productive date query feels nice. The test suite is simpler for it.

**Comment by evilstreak** [comment]
Nice, this looks good! Squash away :)

**Pull request merged**